pkg/proxy/endpointslicecache.go

```diff
@@ -144,6 +144,7 @@ func (cache *EndpointSliceCache) checkoutChanges() map[types.NamespacedName]*end
 		for name, sliceData := range esTracker.pending {
 			if sliceData.remove {
 				delete(esTracker.applied, name)
+				delete(cache.trackerByServiceMap, serviceNN)
```

Comment:
nah, it seems to delete all the service map
```diff
diff --git a/pkg/proxy/endpointslicecache_test.go b/pkg/proxy/endpointslicecache_test.go
index 0281bed2455..54cae55dfb7 100644
--- a/pkg/proxy/endpointslicecache_test.go
+++ b/pkg/proxy/endpointslicecache_test.go
@@ -592,6 +592,13 @@ func TestEndpointSliceCacheClearedCorrectly(t *testing.T) {
                                makeTestEndpointSlice("ns1", "ep2", 2, initEndpointsPorts),
                        },
                },
+               {
+                       name: "two endpoint slices, same service",
+                       currEndpointSlices: []*discovery.EndpointSlice{
+                               makeTestEndpointSlice("ns1", "ep1", 1, initEndpointsPorts),
+                               makeTestEndpointSlice("ns1", "ep1", 2, initEndpointsPorts),
+                       },
+               },
                {
                        name: "two endpoint slices, different namespace",
                        currEndpointSlices: []*discovery.EndpointSlice{
@@ -623,3 +630,41 @@ func TestEndpointSliceCacheClearedCorrectly(t *testing.T) {
                })
        }
 }
+
+func TestSameServiceEndpointSliceCacheClearedCorrectly(t *testing.T) {
+
+       initEndpointsPorts := func(eps *discovery.EndpointSlice) {
+               eps.Endpoints = []discovery.Endpoint{{
+                       Addresses: []string{"1.1.1.1"},
+                       NodeName:  ptr.To(testHostname),
+               }}
+               eps.Ports = []discovery.EndpointPort{{
+                       Name:     ptr.To("n1"),
+                       Port:     ptr.To[int32](11),
+                       Protocol: ptr.To(v1.ProtocolUDP),
+               }}
+       }
+
+       currEndpointSlices := []*discovery.EndpointSlice{
+               makeTestEndpointSlice("ns1", "svc1", 1, initEndpointsPorts),
+               makeTestEndpointSlice("ns1", "svc1", 2, initEndpointsPorts),
+       }
+
+       fp := newFakeProxier(v1.IPv4Protocol, time.Time{})
+       fp.hostname = testHostname
+
+       for _, epSlice := range currEndpointSlices {
+               fp.addEndpointSlice(epSlice)
+       }
+       fp.endpointsMap.Update(fp.endpointsChanges)
+
+       // only delete the first endpoint slice
+       fp.deleteEndpointSlice(currEndpointSlices[0])
+
+       fp.endpointsMap.Update(fp.endpointsChanges)
+
+       if len(fp.endpointsChanges.endpointSliceCache.trackerByServiceMap) != 1 {
+               t.Errorf("expected: endpointSliceCache to have one entrie, got: endpointSliceCache has %d entries", len(fp.endpointsChanges.endpointSliceCache.trackerByServiceMap))
+       }
+
+}
```

we can only delete the serviceNN when there is no endpointslice referencing it, not just one of the slices is deleted ... please add these test cases too

Reply:
> data structure

The creation of the service may take approximately 1 second and a syscall to ipvs may need 50ms.The allocation of a new endpointSliceTracker struct may take only less than 1 ms. If we decide to change this data structure, it may require some additional time for the system to reach a stable state.